jax/interpreters/numpy_eval.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from contextlib import contextmanager
+from typing import Any, Dict, Callable
+
+from jax import config, core
+from jax.interpreters.xla import DeviceArray
+
+np_impl: Dict[core.Primitive, Callable[..., Any]] = {}
+is_tracing = False
+
+class NumpyEvalTrace(core.Trace):
+  def pure(self, x): return x
+  lift = sublift = pure
+
+  def process_primitive(self, primitive, tracers, params):
+    impl = np_impl.get(primitive)
+    if impl is None:
+      raise NotImplementedError(f"No original numpy implementation for "
+                                f"'{primitive}' primitive found.")
+    tracers = [t.copy() if isinstance(t, DeviceArray) else t for t in tracers]
+    return impl(*tracers, **params)
+
+  def process_call(self, primitive, f, tracers, params):
+    return f.call_wrapped(*tracers)
+  process_map = process_call
+
+@contextmanager
+def numpy_eval_context():
+  """
+  Makes JAX code evaluate with numpy as a backend instead of XLA.
+
+  >>> from jax.interpreters.numpy_eval import numpy_eval_context
+  ... with numpy_eval_context():
+  ...   # jax code
+  """
+  assert config.omnistaging_enabled, \
+    "Numpy evaluation requires omnistaging. Set flag JAX_OMNISTAGING=True or " \
+    "call jax.config.enable_omnistaging() before using jax."
+  stack = core.thread_local_state.trace_state.trace_stack
+  prev_dynamic = stack.dynamic
+  stack.dynamic = core.MasterTrace(prev_dynamic.level, NumpyEvalTrace)
+  global is_tracing
+  is_tracing = True
+  try:
+    yield
+  finally:
+    is_tracing = False
+    stack.dynamic = prev_dynamic
+
+def numpy_eval(fun):
```

Comment:
API surface is expensive; let's delete this and just have `numpy_eval_context`.

Reply:
👍  you can still use it as a decorator like `@numpy_eval_context()`

Reply:
Done! Renamed `numpy_eval_context ` to `numpy_eval` to reduce clutter.